### DISCUSSION TITLE
Cannot update stateful material property with its old value:

### DISCUSSION THREAD

**Discussion by steam-simulator** [idaholab/moose#24291] [Q&A Modules: General] [answered]
Hello,
I’m using the `computeQpProperties` virtual function to update my stateful material property but I keep getting a constant value. It seems my stateful material property is not changing. Please, help.

my h file:
```
MaterialProperty<Real>& _b;
const MaterialProperty<Real>& _b_old;

const VariableValue& _sat;

const Real _rm;
```

my c file:
```
InputParameters
ABC::validParams()
{
  InputParameters params = PorousFlowMaterial::validParams();
  params.addRequiredCoupledVar("sat", "saturation");
  params.addParam<Real>("rm", 0.277, " coefficient");
  
  return params;
  }

ABC::ABC(const InputParameters & parameters)
    : PorousFlowMaterial(parameters),
  _b(_nodal_material
                   ? declareProperty<Real>("initial_fracture_aperture_nodal")
                   : declareProperty<Real>("initial_fracture_aperture_qp")),
  _b_old(_nodal_material
                   ? getMaterialPropertyOld<Real>("initial_fracture_aperture_nodal")
                   : getMaterialPropertyOld<Real>("initial_fracture_aperture_qp")),
  _sat (coupledValue("sat")),
  _rm(getParam<Real>("rm")) 
  {
  }


void
ABC::initQpStatefulProperties()
{
 _b[_qp] = 1.0;
}


void
ABC::computeQpProperties()
{
 _b[_qp] = _b_old[_qp] - (_dt * _b_old[_qp] * _sat[_qp] *_rm); 
    _console << "b = " << _b[_qp]<< std::endl;
}
```

**Comment by GiudGiud** [comment, marked as answer]
Hello

What do you get from this print? And from printing _b_old?
Maybe you are at a fixed point?

Guillaume

**Reply by steam-simulator** [reply]
Hello @GiudGiud, I got 1 for both _b and _b_old (i.e., the `initQpStatefulProperties` value for _b ). Pls, any idea why I’m getting only this initial value?

**Reply by GiudGiud** [reply]
how much is sat ?
and _rm ?

if either is 0, then 1 is expected:
b = 1 - ... * 0

**Reply by steam-simulator** [reply]
Hi, sat is 0.18 and rm = 0.277.

**Reply by GiudGiud** [reply]
any idea @jiangwen84

**Reply by lindsayad** [reply]
Is this a nodal material? I would not be all that surprised if the nodal material properties that porous flow has implemented do not work with our stateful material property infrastructure in the framework. @cpgr @WilkAndy can you guys comment on that?

**Reply by steam-simulator** [reply]
Hi @GiudGiud @lindsayad. 
```
    _b[_qp] = _b_old[_qp] - (_dt * _b_old[_qp] * _sat[_qp] *_rm); 
    _console << "b = " << _b[_qp]<< std::endl;
    _console << "bOld = " << _b_old[_qp]<< std::endl;
    _console << "Timestep = " << _dt<< std::endl;
```

It works now. I found the source of the problem after I printed out `_dt`. My simulation does not converge. However, it cuts the size of `_dt `to 0.5, but the actual Time step is still 1; therefore, `_b_old` doesn’t get updated even though `_b` changes.  

```
b = 0.95014
bOld = 1
Timestep = 1
b = 0.95014
bOld = 1
Timestep = 1
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 2
 Solve Did NOT Converge!
  Finished Solving                                                                       [  8.98 s] [    3 MB]
Aborting as solve did not converge

Solve failed, cutting timestep.

Time Step 1, time = 0.5, dt = 0.5
b = 0.97507
bOld = 1
Timestep = 0.5
b = 0.97507
bOld = 1
Timestep = 0.5
```

**Reply by cpgr** [reply]
The nodal materials do work with the stateful property infrastructure and we use it extensively. It's just crazy enough to work!

It looks like it will work once your solution converges and the current b gets propagated to bOld.

**Reply by lindsayad** [reply]
Thanks for the info @cpgr !
